[AOTInductor] Reuse generated kernels between constant graph and main graph

muchulee8 opened the pull request:
Summary: We copy the src_to_kernel from constant graph to main graph so that we could avoid generating duplicating kernels. And pass throught the name counter such that no duplicated names will be generated.

Test Plan: Included in commit

Differential Revision: D54706767




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @aakhundov @ColinPeppler @amjames @desertfire @chauhang

chenyang78 reviewed (commented):
Please fix the lint issue. Otherwise, LGTM. Thanks!

aakhundov commented on torch/_inductor/graph.py:
```diff
@@ -1118,21 +1118,26 @@ def init_wrapper_code(self):
+            assert wrapper_code_gen_cls is not None, f"Device {device_type} not supported"
+            self.wrapper_code = wrapper_code_gen_cls()
+
+        if self.const_module:
+            # If we have const module, we could reuse the kernels
+            # This could avoid duplication and save time on doing recompilation (if Triton.)
+            self.wrapper_code._names_iter = self.const_module.wrapper_code._names_iter
+            self.wrapper_code.src_to_kernel = self.const_module.wrapper_code.src_to_kernel
```
Is this relevant for the C++ wrappers?

muchulee8 committed:
[AOTInductor] Reuse generated kernels between constant graph and main graph (#121564)

Summary:

We copy the src_to_kernel from constant graph to main graph so that we could avoid generating duplicating kernels. And pass throught the name counter such that no duplicated names will be generated.

Test Plan: Included in commit

Differential Revision: D54706767

huydhn commented:
@muchulee8 One of the test `inductor/test_aot_inductor.py::AOTInductorTestABICompatibleCuda::test_duplicate_constant_folding_abi_compatible_cuda` is failing on ROCm https://hud.pytorch.org/pytorch/pytorch/commit/767643301209fbe270dbe170886111917f836051.  Could you help take a look?  Adding `ciflow/rocm` on the PR will trigger the job
